README

         set6x86 / get6x86 / 6x86_reg
         ----------------------------

This document has been slightly modified from its original form to
accomodate the OS/2 specific implementation of the program.


Please read through this file before attempting to use this program.
It might save you a few crashes.


Description
-----------

Set6x86 and get6x86 are two programs that allow you to poke and peek into
the Cyrix/IBM/SGS-Thomson 6x86 CPU's configuration registers. 6x86_reg will
display register contents in human-readable format. The short source
test_bug is a demo program for the 6x86 "coma" bug.


Cyrix CPUs
----------

Cyrix 6x86 CPU's are Intel Pentium-like performance CPUs, but at a lower
cost than their Intel parts. Cyrix has adopted a CPU rating that refers to
"equivalent Pentium performance in Windows applications" for their CPUs.

A Cyrix-6x86 P133+ for example is clocked at only 110 MHz, although it
performs as good as a Pentium 133 (hence the "P133+" rating). The "+" is a
marketing gadget: it is supposed to point out that it's even a little better
than a P133 on the Winstone benchmark.

6x86 CPUs come at a significantly lower price than their equivalent Intel
cousins. The only downsides are:

   - a lack of support from several motherboard vendors and software vendors
     (most CPU identification programs, Linux 2.0's /proc/cpuinfo included,
     report a 486 CPU instead of a Cyrix 6x86). A short kernel patch
     is available from http://www.tux.org/~balsa/linux/cyrix that
     solves this kernel shortcoming.

   - Floating point performance is not as good as a Pentium. Depending
     on the benchmark, 15 to 50% lower FPU performance can be expected.

   - This baby needs HUGE amounts of power. People used to make fun of
     the newest Digital chip, the Alpha, a few years ago, because it
     needed over 20 Watts of power. Well, this one too. Depending on
     the clock speed, up to 24 Watts of power is dissipated from the
     package!

     Anyone familiar with hardware design can tell you that keeping
     that under control is not a simple task. That is also the main
     reason why so many people flush newsgroups with messages that the
     6x86 CPU's are bad, don't run Win95, crash all the time,...

     Do you really think those companies would even _risk_ making a CPU
     that is not as compatible as possible? The only thing they didn't
     replicate was the ominous Pentium FPU bug ;-)

          On the other hand Cyrix 6x86 CPUs suffer from another bug, called
          the 6x86 "coma" bug. The short program test_bug included in this
          package is an example of how this bug affects 6x86 CPUs.  If you
          want to see the coma bug in action, simply compile and run it.

     The _real_ reason for crashes and system hangs in most cases is
     either a temperature problem, or a digital timing problem (the
     latter being caused by the former in many cases). Poorly cooled
     CPU's tend to go beyond their specified ratings. Most CPUs are
     only guaranteed to work up to 70 degrees Celsius, and with 24
     Watts to sweat out from such a small surface, that's _very_ easily
     exceeded.


Why this program?
-----------------

For fun, mostly.

But also for some pretty darn good reasons. Most motherboards (mine, a
Chaintech board based on the Intel Triton II chipset) support 6x86 CPUs, but
not as well-tested and well-set-up as Pentium CPUs.

The 6x86 has a set of configuration registers that allows one to do a lot of
things, like specifying non-cacheable memory areas (important for e.g.
graphics cards), I/O delays, cache write policy (WB/WT), and also to enable
an automatic standby mode where a CPU "halt" instruction cuts down power by
a factor of 70 (from 5.8 Amps to 83 milli-Amps for the P133+).

Since most computers do nothing most of the time, this would be useful to
enable.

Linux uses the halt instruction when it has nothing better to do, so
enabling this feature will _not_ hurt performance: the CPU controls the
powering-down of specific internal parts by hardware-control, so there's no
software intervention needed for explicitly re-enabling a powered-down part.

However, some Linux users have reported that their systems would hang if
this feature was enabled and some bus-mastering Ethernet cards were used, in
some specific circumstances and only with some motherboards/chipset/BIOS
combinations! So even this seemingly inocuous feature should be tested on
individual systems.


Benefits?
---------

Apart from the obvious environmental reasons, there's also the temperature
problem that is reduced substantially.

If you're on a P133+ and programming like hell, or reading mail, or most of
the other things you do in Linux (or any other OS for that matter), the CPU
is doing nothing 99% of the time (just take an occasional look at the system
load meter).

When not set for automatic power-down, the CPU will _always_ draw its full
24 Watts of power. With this feature enabled, power consumption drops to
near-zero most of the time.

While I'm typing this junk, my CPU is (again) 99% idle, so now the CPU uses
only 0.3 Watts instead. That's one _hundreth_ of its full juice.

No need to say that most thermal problems are history.

In fact, I've been having _lots_ of spurious core dumps and "signal 11"
messages from GCC on my system, and I always suspected temperature problems.
With auto-suspend mode on, I haven't had a single core dump since (but I
haven't done a lot of testing either).

This convinced me that I'll have to provide better cooling for the CPU.


DANGER!!!
---------

Of course, you're fiddling with the guts of your CPU, while it's running. So
don't blame me if you accidently set the wrong register and your machine
hangs.

In other words, let's use the Paul Gortmaker (?) disclaimer:

   "If it breaks, you get to keep the pieces"


LITERATURE
----------

This program is completely useless without the proper docs from Cyrix. I
explicitly didn't include any description of the 6x86 config registers
because I didn't want to risk distributing faulty information.

Cyrix's WWW site (www.cyrix.com) and their FTP site (ftp.cyrix.com) contain
all the stuff you need. For ftp, all the files are in the "6x86" directory.

You need _at least_

   6X-ABDB.PDF   (361748 bytes)

which is the abbreviated data book, and contains a description of all the
registers you can tamper with (staring at page 18).

The full data book is

   6X-DBOOK.PDF  (2039298 bytes)

For the 5x86, the Cyrix FTP/WWW site may contain equivalent doumentation
(not checked).

Here's a list of WWW sites known to carry information on 6x86 CPUs:

  http://www.cyrix.com
  http://www.chips.ibm.com/products/x86/
  http://www.tux.org/~balsa/linux/cyrix/index.html
  http://www.alternativecpu.com


EXAMPLE
-------

The power-saving option described above is controlled by register 0xC2, bit
3 of the configuration registers.  The register is specified with the "-r"
option.  Thus,

   get6x86 -r 0xC2

Will display (on my system -- "your mileage my vary"):

   Cyrix 6x86 config register, index 194 (=0xc2) contains 128 (=0x80 =b10000000)

Note that bit 3 is set to zero, so the auto-suspend mode is disabled.

Typing

   set6x86 -r 0xC2 -d 0x88

will set bit #3 of that register, thus enabling the auto-suspend mode.  Note
that the "-d" option assigns a raw data value to the register.


Simpler is to use the "-s" or "-c" options. These use the data byte on the
command line as a set/clear mask instead of raw data to program into the
register.

So the above can be made much more simple by just typing

   set6x86 -r 0xC2 -s 0x08

This will read register 0xC2, and "set" bits using the mask given (0x08).

The "-s" does a logical "OR" of the original register data and the mask
data, and programs it back into the register, whilst the "-c" option will do
a logical "AND" with the inverse of the mask (0xF7 in this case).


The included cyrix.cmd file enables a few performance-related thingies on the
Cyrix 6x86 CPU. Copy it to the /etc/rc.d directory, and call it from e.g.
rc.local.

It is also instructive on how to use both the tool, and the Cyrix 6x86
registers. It does a bit of everything.


6x86_reg
--------

This little tool dumps most relevant 6x86 registers in human-readable
format. It will also correctly identify the 6x86 model and revision, report
bogoMIPS ratings, and dump the more complex ARRs (Address Region Registers).
These are the registers where the different memory regions are set up and
how they are handled (e.g. if they're cached or if any other special
function is enabled for them).

Running it may reveal something like the output below:

* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
6x86 (Classic/L/MX) Register Dump utility

6x86 DIR0: 0x52   2.5X core/bus clock ratio
     DIR1: 0x4 6x86MX Rev. 1.4

Wait a moment...
Calculated BogoMIPS: 185.12
Kernel BogoMIPS:  186.78

6x86 CCR0: 0x2 NC1 set (address region 640Kb-1Mb non-cacheable)
     CCR1: 0x92   NO_LOCK set
     CCR2: 0x88   SUSP_HLT set (low power suspend mode enabled)
     CCR3: 0x10
     CCR4: 0x87   no I/O recovery time
     CCR5: 0x21   allocate cache lines on write misses

6x86 Address Region Register dump:
  ARR0: address = 0xA0000 , size = 128 KB
    RCR = 0x9 : not cached, write gathering
  ARR1: address = 0xC0000 , size = 256 KB
    RCR = 0x1 : not cached
  ARR2: disabled
  ARR3: address = 0xA8000 , size = 32 KB
    RCR = 0x9 : not cached, write gathering
  ARR4: disabled
  ARR5: disabled
  ARR6: address = 0xE0000000 , size = 4 MB
    RCR = 0x9 : not cached, write gathering
  ARR7: address = 0x0 , size = 32 MB
    RCR = 0x9 : cached, write gathering
* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *

This saves you a lot of fingering through the 6x86 data book :-)


BogoMIPS
--------

First of all you must understand that bogoMIPS are _not_ meant as a measure
of performance, but rather as an indicator that the 6x86 registers and the
BIOS settings for the chipset on your motherboard are correctly set.

That said, we can look at the three possible outcomes of running 6x86_reg:

1) The calculated bogoMIPS and the kernel bogoMIPS reading from
/proc/cpuinfo are approx. equal (+/- 1 bogoMIPS). This is the normal case on
a "quiet" system.

Note that on 6x86 family CPUs, bogoMIPS are approximately equal to the CPU
core clock rate.

For example:
   - a 6x86 PR166 should have a bogoMIPS rating of approx. 133, since
          its core will run at 133MHz (2 x 66.66MHz).

   - a 6x86MX PR233 at 187.5MHz (2.5 x 75MHz) should have a bogoMIPS
          rating of approx. 187.5.

2) The calculated bogoMIPS is much higher (> 10-15%) than the kernel
bogoMIPS. This indicates that your BIOS is not setting the 6x86 registers or
the chipset parameters for the L2 cache correctly, and that set6x86 has
remedied to this. Adjust your BIOS CMOS settings and/or upgrade the BIOS
flash EEPROM.

3) The calculated bogoMIPS is lower than the /proc/cpuinfo kernel reading:
your system load is impairing the bogoMIPS calculation done by 6x86_reg.


Fun?
----

You might even be able to _hear_ the difference!

Most computer power supplies only compensate higher power demands on the +5V
rail, resulting in an increase on the +12V rail when +5V power demand goes
up.

The CPU draws from the +5V rail, but the CPU fan uses +12V, so when the CPU
draws more power (due to CPU load), the CPU cooling fan (which is delivered
with the CPU by Cyrix) will run a little faster with higher supply voltages,
so the significant power drop when the CPU drops into power-suspend mode
might cause a noticeable change in the pitch of the buzzing fan sound!

On my system, I can now HEAR how busy it is...


The 6x86 Coma bug
-----------------

Compile and run the short bug demo program "test_bug.c". Type Control-C to
try and regain control: nothing happens, right? Your 6x86 CPU is in "coma"
mode. :-(

The 6x86 CPU effectively gets locked in an infinite loop, executing
back-to-back locked cycles and consequently unable to service any
interrupts, including attempts by root to kill the process. This is by all
means a serious security flaw in a multiuser OS like Linux.

This bug affects all 6x86 family CPUs.

Solution: enable the NO_LOCK bit with set6x86.


Support
-------

None. Your own intelligence and common sense should be enough.


DOS and Linux
-------------

DOS support was removed starting at version 1.4.  DOS, not being a _real_
Operating System, does busy-waiting when it's idle instead of executing a
"halt" instruction, and thus the CPU can't know that it should be powering
down. So it'll still run as hot as under a full load.

Windows 3.1 and '95 don't as well. Blame MicroBrains for that. NT? Don't know.

The Linux version can be downloaded from:

   ftp://tsx-11.mit.edu/pub/linux/sources/sbin

The Linux version works great though.  (It's a real operating system)


Author
-------

Koen Gadeyne, koen.gadeyne@barco.com
(please contact the maintainer below first)


New maintainer for the general/Linux version
--------------------------------------------

Thanks Koen! I'll do my best with this "hot potato" :-)
Andrew D. Balsa, andrewbalsa@usa.net


OS/2 Modifications/Port:
-----------------------

Greg Kondrasuk, kondrag@geocities.com
http://www.geocities.com/SiliconValley/Vista/6434/


CHANGES
14DEC97 -- version 1.5

- 6x86_reg now also displays two bogoMIPS readings: the kernel-calculated
  bogomips from/proc/cpuinfo, and an independently-calculated bogomips reading.

- Note that bogoMIPS are *NOT* a benchmark or a test for CPU performance; they
  are just a very good indicator of whether the BIOS has correctly configured
  the 6x86 registers and the L2 cache/chipset timing parameters.


03DEC97 -- version 1.4

- 6x86_reg replacement for 6x86_arr (now dumps relevant 6x86 registers
  in human-readable format).
- DOS support was dropped.
- A short, simple kernel patch is included.
- With Mike Jagdis' permission, a new documentation file Doc-6x86.txt was
  included with this release of set6x86. It does not replace the relevant
  Cyrix and IBM Application Notes and Data Books, which one should read
  anyway to get a better understanding of the 6x86 family of CPUs.
- New maintainer: Andrew D. Balsa <andrewbalsa@usa.net>.


14SEP97 -- version 1.3

- 6x86_arr added (dumps Address Region Registers in human-readable format)


29OCT96 -- version 1.2

- "-c" and "-s" options for bit-setting and bit-masking.
- silent mode changed so that get6x86reg outputs HEX-only
  (otherwise it's pretty useless).
- silent mode renamed to "-p" (pipeable output)
  for compatibility with "setVGAreg".
- fatal command line parameter errors now exit instead of
  dumping core...
- DOS version now uses the DOS cross-compiler from DJGPP.


10OCT96 -- version 1.1

- some documentation updates.
- DOS version added (DJGPP compiled) + docs
- silent mode (no output).


05AUG96 -- version 1.0

- initial release. Hacked up from "setVGAreg", which 
  is part of SVGATextMode.


test_bug.c
/* This program is DANGEROUS: it will lock (freeze) your system */
/* Read all the available documentation before executing it */

static unsigned char c[4] = {0x36, 0x78, 0x38, 0x36};
main()
{
asm ("movl	$c, %ebx\n\t"
"again:	xchgl	(%ebx), %eax\n\t"
	"movl	%eax, %edx\n\t"
	"jmp	again\n\t");
}
	

set6x86.c
/*  get/set6x86: a tool for changing Cyrix 6x86 configuration registers
 *
 *  Copyright (C) 1996  Koen Gadeyne
 *  Modifications for OS/2 port by Greg Kondrasuk, 1997.
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */


/***
 *** This is just a hacking tool! Use at your own risk. It was NOT intended to be
 *** idiot proof! If you don't understand all this, then don't bother trying to use it.
 ***
 ***/

#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <stdarg.h>
#include <stdlib.h>
#include <string.h>
#include <sys/hw.h>

#define outb(data,port) _outp32(port,data)
#define inb(port) _inp32(port)
#define ioperm(x,y,z) (0)
#define VERSION 1.5

#ifndef TRUE
#  define TRUE (1)
#endif
#ifndef FALSE
#  define FALSE (0)
#endif

typedef int bool;

char *CommandName;


//////////////////////////////////////////////////////////////////////////
// Function: get_IO_range
//////////////////////////////////////////////////////////////////////////
void get_IO_range(int start, int len)
{
   if (ioperm(start, len, 1) != 0) {
      perror("I/O Permissions");
      fprintf(stderr,"Cannot get I/O permissions for hardware address range 0x%x-0x%x.\n\
             You must be superuser, or the program must be setuid root!\n", start, start+len-1);

      exit(1);
   }
}

//////////////////////////////////////////////////////////////////////////
// Function: check_int_range
//////////////////////////////////////////////////////////////////////////
void check_int_range(int cvalue, int lmin, int lmax, char* descstr)
{
   if (cvalue<lmin || cvalue>lmax) {
      fprintf(stderr,"%s = %d (0x%x) out of range [%d..%d]!\n", descstr, cvalue, cvalue, lmin, lmax);

      exit(1);
   }
}

//////////////////////////////////////////////////////////////////////////
// Function: getint
// Description: Converts the int in 'instring' into an integer. Must be
//              within specified limits 'lmin' and 'lmax'. 'descrstring'
//              contains a description of the number to be parsed, used
//              in the error message.
//////////////////////////////////////////////////////////////////////////
int getint(char* instring, char* descrstring, int lmin, int lmax)
{
   char** errstr=NULL;
   int readint;
   readint = strtol(instring,errstr,0);

   if (errstr) {
      fprintf(stderr,"Illegal character '%s' in %s: '%s'\n", *errstr, descrstring, instring);
      exit(1);
   }

   check_int_range(readint, lmin, lmax, descrstring);

   return(readint);
}

//////////////////////////////////////////////////////////////////////////
// Function: int_to_bin
//////////////////////////////////////////////////////////////////////////
char* int_to_bin(int num, int bits)
{
   static char binstr[sizeof(long int)+1];
   int i;

   for (i=0; i<bits; i++)
      binstr[i] = 0x30; // char '0'

   binstr[bits] = 0x00;

   for (i=0; i<bits; i++)
      binstr[bits-1-i] += ((num >> i) & 0x00000001);

   return(binstr);
}

//////////////////////////////////////////////////////////////////////////
// Function: usage
// Description: displays program usage information.
//////////////////////////////////////////////////////////////////////////
void usage(bool setreg)
{
   printf("Inside the usage function.\n");
   printf("%s %.1f (c) 1996 Koen Gadeyne.\n", CommandName, VERSION);
   printf("Modified for OS/2 by Greg Kondrasuk.");
   printf("\n"
   "  Usage: %s [-hp] [-r register_index] %s\n\n", CommandName, (setreg) ? "[-d data] | [-sc bitmask]" : "");
   printf("  Options: -h  print usage information\n"
   "%s"\
   "\n"
   "  register_index: An index to the specified Cyrix 6x86 config register,\n"
   "                  In decimal (e.g. '24'), hex ('0x18') or octal ('030').\n"
   "\n"
   "%s\n",
   (setreg) ? \
   "           -p  silent mode: don't spread the news (no output)\n"\
   "           -r  `register_index' specifies the register to use\n"\
   "           -d  `data' is a raw value to program into the register\n"\
   "           -s  `data' is a bitmask for setting bits in the register\n"\
   "           -c  `data' is a bitmask for clearing bits in the register\n"\
   : \
   "           -p  pipeable output: only hex register contents are printed\n",\
   (setreg) ? \
   "  data: for the -d option, the raw data to program into the specified\n"\
   "        register (dec|hex|oct), or the bitmask for the -s and -c options.\n"\
   : \
   "");
}


/***********************************************************************************************************/

int main (int argc, char* argv[])
{
   int c;
   int tmpbyte=0;
   int regnum=0;
   bool setreg = FALSE;
   bool silent = FALSE;
   bool arg_is_setmask=FALSE;
   bool arg_is_clearmask=FALSE;
   char* commandfilename;
   int data=0;


   // See what action is required: read or write

   CommandName = argv[0];
   commandfilename = strrchr(CommandName, '\\');

   if (commandfilename)
      commandfilename++;
   else
      commandfilename = CommandName;

   setreg = (!strnicmp(commandfilename,"set",3));



   // command-line argument parsing

   while ((c = getopt (argc, argv, "hpr:d:s:c:")) != EOF) {
      switch (c) {

         case 'h': usage(setreg);
                   exit(0);
                   break;
         case 'p': silent = TRUE;
                   break;
         case 'r': regnum = getint( optarg, "register index", 0, 255 );
                   break;
         case 'd': data = getint( optarg, "register data", 0, 255 );
                   break;
         case 's': arg_is_setmask = TRUE;
                   data = getint( optarg, "register data", 0, 255 );
                   break;
         case 'c': arg_is_clearmask = TRUE;
                   data = getint( optarg, "register data", 0, 255 );
                   break;
         case '?': usage(setreg);
                   exit(1);
                   break;
         default:  fprintf(stderr,"getopt returned unknown token '%c'.\n",c);
                   exit(1);
      }
   }

   get_IO_range(0x22, 2);

   if (setreg) {
      if (arg_is_clearmask || arg_is_setmask) {

         outb(regnum, 0x22); tmpbyte = inb(0x23);  // read original data

         if (arg_is_clearmask)
            data = tmpbyte & ~data;
         if (arg_is_setmask)
            data = tmpbyte | data;
      }

      outb(regnum, 0x22); outb(data, 0x23);          // write data
   }

   outb(regnum, 0x22); tmpbyte = inb(0x23);         // read back data

   if (!silent)
       printf("Cyrix 6x86 config register, index %d (=0x%x) contains %d (=0x%02x =b%s)\n",
               regnum, regnum, tmpbyte, tmpbyte, int_to_bin(tmpbyte,8));

   else if (!setreg)                // "getreg" needs to output _something_
           printf("0x%x\n", tmpbyte);


   return(0);

}  // end main


6x86_reg.c
/*  6x86_reg: a tool for displaying Cyrix 6x86 configuration registers
 *  and ARRs in human-readable format.
 *
 *  Copyright (C) 1996, 1997  Koen Gadeyne
 *  Modifications for OS/2 port by Greg Kondrasuk, 1998.
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 *  Updated 12/97 by Andrew D. Balsa
 *  (see http://www.tux.org/~balsa/linux/cyrix)
 *
 ******************************************************************************
 *
 *  Usage: no parameters! Just typing 6x86_reg at the prompt will dump all 6x86(L)
 *  documented registers to the console. Now of course if you run it on another
 *  processor model don't expect any meaningful results...
 */
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <stdarg.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
//#include <asm/io.h>  not needed by OS/2

// start of OS/2 section added
#include <sys/hw.h>
#define outb(data,port) _outp32(port,data)
#define inb(port) _inp32(port)
#define ioperm(x,y,z) (0)
#define VERSION 1.5
// end of OS/2 section

typedef int bool;
#ifndef TRUE
#  define TRUE (1)
#endif
#ifndef FALSE
#  define FALSE (0)
#endif
char *CommandName;
char info[256];
void get_IO_range(int start, int len)
{
  if (ioperm(start, len, 1) != 0)
  {
    perror("I/O Permissions");
    fprintf(stderr,"Cannot get I/O permissions for hardware address range 0x%x-0x%x.\n\
             You must be superuser, or the program must be setuid root!\n", start, start+len-1);
    exit(1);
  }
}
char* int_to_bin(int num, int bits)
{
  static char binstr[sizeof(long int)+1];
  int i;
  for (i=0; i<bits; i++) binstr[i] = 0x30; /* char '0' */
  binstr[bits] = 0x00;
  for (i=0; i<bits; i++) binstr[bits-1-i] += ((num >> i) & 0x00000001);
  return(binstr);
}
char *size_str(int kb)
{
  static char size[20];
  if (kb < 1024) sprintf(size, "%d KB", kb);
  else if (kb < 1024*1024) sprintf(size, "%d MB", kb >> 10);
  else sprintf(size, "%d GB", kb >> 20);
  return size;
}
/* the original bogomips code from the Linux kernel */
static __inline__ void delay(int loops)
{
  __asm__(".align 2,0x90\n1:\tdecl %0\n\tjns 1b": :"a" (loops):"ax");
}
/* a short routine to read /proc/cpuinfo */
void read_cpuinfo(void) {
  FILE *in;
  char things[256];
  extern char info[256];
  if ((in=fopen("/proc/cpuinfo","r"))==NULL)
      strcpy(info,"unknown");
  else {
   while (strcmp(things,"bogomips"))
     fscanf(in, "%s", things);   /* skip to the bogomips line */
   fscanf(in, "%s", things);  /* skip the ":" */
   fscanf(in, "%s", info);    /* now this should be the bogomips */
   fclose(in);
  }
  return;
}
/* Register index defines */
#define ARR_BASE 0xC4
#define RCR_BASE 0xDC
#define CCR0 0xC0
#define CCR1 0xC1
#define CCR2 0xC2
#define CCR3 0xC3
#define CCR4 0xE8
#define CCR5 0xE9
#define DIR0 0xFE
#define DIR1 0xFF
/*****************************************************************************/
/* This was a nice short routine before I began messing with it
 * It still has a very linear structure, but the registers on the 6x86 line
 * of CPUs are so irregular that the program got a little bit messy.
 * - Andrew
 */
/*****************************************************************************/
int main (int argc, char* argv[])
{
  int i;
  unsigned int addr, size, rcr, ccr, dir;
  int mapen;
  unsigned long loops_per_sec = 1;
  unsigned long ticks;
/*
 * start doing something useful
 */
   get_IO_range(0x22, 2);
   printf("\n6x86 (Classic/L/MX) Register Dump utility\n\n");
  /* MAPEN **************************************************************/
  outb(0xC3, 0x22); mapen = inb(0x23);
  outb(0xC3, 0x22); outb(mapen | 0x10, 0x23);
  /* DIR0 ***************************************************************/
  outb(DIR0, 0x22); dir = inb(0x23);
  printf("6x86 DIR0: 0x%x\t", dir);
  /* core/bus clock ratio analysis **************************************/
  if ((dir & 0x70) == 0x50)   /* 6x86MX */
  switch (dir & 0x07) {
   case 0x01:
      { printf("2X core/bus clock ratio");
        break;
      }
   case 0x02:
      { printf("2.5X core/bus clock ratio");
        break;
      }
   case 0x03:
      { printf("3X core/bus clock ratio");
        break;
      }
   case 0x04:
      { printf("3.5X core/bus clock ratio");
        break;
      }
  }
  else            /* 6x86(L) */
  switch (dir & 0x05) {
   case 0x00:
      { printf("1X core/bus clock ratio");
        break;
      }
   case 0x01:     /* practically the only one used */
      { printf("2X core/bus clock ratio");
        break;
      }
   case 0x05:
      { printf("3X core/bus clock ratio");
        break;
      }
   case 0x04:
      { printf("4X core/bus clock ratio");
        break;
      }
  }
  printf("\n");
  /* DIR1 ***************************************************************/
  outb(DIR1, 0x22); dir = inb(0x23);
  printf("     DIR1: 0x%x\t", dir);
  /* Processor identification *******************************************/
  switch (dir) {
   case 0x14:
      { printf("6x86 Rev. 2.4");
        break;
      }
   case 0x15:
      { printf("6x86 Rev. 2.5");
        break;
      }
   case 0x16:
      { printf("6x86 Rev. 2.6");
        break;
      }
   case 0x17:
      { printf("6x86 Rev. 2.7 or 3.7");
        break;
      }
   case 0x22:
      { printf("6x86L Rev. 4.2");
        break;
      }
   case 0x03:
      { printf("6x86MX Rev. 1.3");
        break;
      }
   case 0x04:
      { printf("6x86MX Rev. 1.4");
        break;
      }
  }
  /* BogoMIPS calculation ***********************************************/
  printf("\n\nWait a moment...\n");
  /* Now let's take a look at the two bogomips readings */
  /* First we calculate the bogoMIPS reading using exactly
   * the same code as the Linux kernel
   */
   while ((loops_per_sec <<= 1)) {
    ticks = clock();
    delay(loops_per_sec);
    ticks = clock() - ticks;
    if (ticks >= CLOCKS_PER_SEC) {
      loops_per_sec = (loops_per_sec / ticks) * CLOCKS_PER_SEC;
      printf("Calculated BogoMIPS:\t%lu.%02lu\n",
        loops_per_sec/500000,
        (loops_per_sec/5000) % 100
        );
      break;
      }
  }
  /* and now the kernel bogoMIPS, calculated at boot time and
   * reported by /proc/cpuinfo.
   */
  read_cpuinfo();
  printf("Kernel BogoMIPS:\t%s\n", info);
  printf("\n");
  /* CCR ****************************************************************/
  outb(CCR0, 0x22); ccr = inb(0x23);
  printf("6x86 CCR0: 0x%x\t", ccr);
  if (ccr & 0x02) printf("NC1 set (address region 640Kb-1Mb non-cacheable)\n");
  else printf("NC1 reset (address region 640Kb-1Mb cacheable)\n");
  outb(CCR1, 0x22); ccr = inb(0x23);
  printf("     CCR1: 0x%x\t", ccr);
  if (ccr & 0x10) printf("NO_LOCK set\n");
  else printf("NO_LOCK reset\n");
  outb(CCR2, 0x22); ccr = inb(0x23);
  printf("     CCR2: 0x%x\t", ccr);
  if (ccr & 0x08) printf("SUSP_HLT set (low power suspend mode enabled)\n");
  else printf("SUSP_HLT reset (low power suspend mode disabled)\n");
  outb(CCR3, 0x22); ccr = inb(0x23);
  printf("     CCR3: 0x%x\n", ccr);
  outb(CCR4, 0x22); ccr = inb(0x23);
  printf("     CCR4: 0x%x\t", ccr);
  if (dir & 0xf0)    /* if this is a 6x86 or 6x86L */
    if (ccr & 0x10) printf("DTE cache enabled, ");
    else printf("DTE cache disabled, ");
  switch (ccr & 0x07) {
   case 0x00:
      { printf("1 clock cycle I/O recovery time");
        break;
      }
   case 0x01:
      { printf("2 clock cycles I/O recovery time");
        break;
      }
   case 0x02:
      { printf("4 clock cycles I/O recovery time");
        break;
      }
   case 0x03:
      { printf("8 clock cycles I/O recovery time");
        break;
      }
   case 0x04:
      { printf("16 clock cycles I/O recovery time");
        break;
      }
   case 0x05:
      { printf("32 clock cycles I/O recovery time");
        break;
      }
   case 0x06:
      { printf("64 clock cycles I/O recovery time");
        break;
      }
   case 0x07:
      { printf("no I/O recovery time");
        break;
      }
  }
  printf("\n");
  outb(CCR5, 0x22); ccr = inb(0x23);
  printf("     CCR5: 0x%x\t", ccr);
  if (dir & 0xf0)    /* if this is a 6x86 or 6x86L */
    if (ccr & 0x02) printf("slow LOOP enabled, ");
    else printf("slow LOOP disabled, ");
  if (ccr & 0x01) printf("allocate cache lines on write misses\n\n");
  else printf("default cache line allocation policy\n\n");
  /* ARR ****************************************************************/
  printf("6x86 Address Region Register dump:\n");
  for (i=0; i<8; i++)
  {
    outb(ARR_BASE + i*3, 0x22); addr = inb(0x23) << 24;
    outb(ARR_BASE + i*3 + 1, 0x22); addr |= inb(0x23) << 16;
    outb(ARR_BASE + i*3 + 2, 0x22); size = inb(0x23);
    addr |= (size & 0xF0) << 8;
    size &= 0x0F;
    printf("  ARR%d: ", i);
    if (size == 0)
    {
      printf("disabled\n");
      continue;
    }
    else
      printf("address = 0x%X , size = ", addr);
    if (i<7)
    {
      if (size == 15) printf("%s\n", size_str(4*1024*1024));
      else printf("%s\n", size_str(2 << size));
    }
    else
    {
      printf("%s\n", size_str(128 << size));
    }
    /* RCR **************************************************************/
    outb(RCR_BASE + i, 0x22); rcr = inb(0x23);
    printf("    RCR = 0x%X : ", rcr);
    if (i != 7)
      printf("%scached",  (rcr & 0x01) ? "not " : "");
    else
      printf("%scached",  (rcr & 0x01) ? "" : "not ");
    if (dir & 0xf0)     /* if this is a 6x86 or 6x86L */
      if (rcr & 0x2) printf(", weak write ordering");
    if (rcr & 0x4) printf(", weak locking");
    if (rcr & 0x8) printf(", write gathering");
    if (rcr & 0x10) printf(", write through");
    if (rcr & 0x20) printf(", no LBA#");
    printf("\n");
  }
  /* Disable MAPEN ******************************************************/
  outb(0xC3, 0x22); outb(mapen, 0x23);
  return(0);
}
